units/migrate_volume_job.go

```diff
@@ -0,0 +1,298 @@
+	defer j.finishJob(ctx)
+
+	if err := j.populateIfUnset(); err != nil {
+		j.AddRetryableError(err)
+		return
+	}
+
+	if err := j.volume.SetMigrating(true); err != nil {
+		j.AddRetryableError(err)
+		return
+	}
+
+	if !j.InitialHostStopped {
+		initialHostStopped, hasErr := j.stopInitialHost(ctx)
+		if hasErr {
+			return
+		}
+		if !initialHostStopped {
+			j.AddRetryableError(errors.Errorf("initial host '%s' not yet stopped", j.InitialHostID))
+			return
+		}
+	}
+
+	if !j.VolumeDetached {
+		// Unmount volume from initial host
+		if err := j.initialHost.UnsetHomeVolume(); err != nil {
+			j.AddError(errors.Wrapf(err, "unsetting home volume '%s' from host '%s'", j.VolumeID, j.InitialHostID))
+			return
+		}
+
+		if _, err := cloud.DetachVolume(ctx, j.VolumeID); err != nil {
+			j.AddError(errors.Wrapf(err, "detaching volume '%s'", j.VolumeID))
+			return
+		}
+		j.VolumeDetached = true
+
+		// Update in-memory volume
+		volume, err := j.getVolume()
+		if err != nil {
+			j.AddError(err)
+			return
+		}
+		j.volume = volume
+	}
+
+	// Avoid recreating new host on retry
+	if j.NewHostID == "" {
+		if hasErr := j.startNewHost(ctx); hasErr {
+			return
+		}
+	}
+
+	// Terminate initial host
+	if !j.InitialHostTerminated {
+		ts := utility.RoundPartOfMinute(1).Format(TSFormat)
+		terminateJob := NewSpawnHostTerminationJob(j.initialHost, evergreen.User, ts)
+		if err := j.env.RemoteQueue().Put(ctx, terminateJob); err != nil {
+			j.AddRetryableError(errors.Wrapf(err, "enqueueing initial host termination job '%s'", terminateJob.ID()))
+			return
+		}
+		j.InitialHostTerminated = true
+	}
+}
+
+// stopInitialHost inspects the initial host's status and stops it if the host is still running.
+// It returns a boolean indicating whether the host has stopped and a boolean indicating whether an error was added to the job.
+func (j *volumeMigrationJob) stopInitialHost(ctx context.Context) (bool, bool) {
+	// Stop host if this operation has not been attempted yet
+	if j.SpawnhostStopJobID == "" {
+		ts := utility.RoundPartOfMinute(1).Format(TSFormat)
+		stopJob := NewSpawnhostStopJob(j.initialHost, evergreen.User, ts)
+		err := j.env.RemoteQueue().Put(ctx, stopJob)
+		if err != nil {
+			j.AddRetryableError(err)
+			return false, true
+		}
+		grip.Info(message.Fields{
+			"message": "stopping initial host",
+			"host_id": j.InitialHostID,
+		})
+		j.SpawnhostStopJobID = stopJob.ID()
+		return false, false
+	}
+
+	// Mark host as stopped
+	if j.initialHost.Status == evergreen.HostStopped {
+		grip.Info(message.Fields{
+			"message": "initial host stopped",
+			"host_id": j.InitialHostID,
+		})
+		j.InitialHostStopped = true
+		return true, false
+	}
+
+	// If host is not stopped, verify whether the job is still in progress and error if it completed without stopping
+	stopJob, foundJob := j.env.RemoteQueue().Get(ctx, j.SpawnhostStopJobID)
+	if !foundJob {
+		j.AddRetryableError(errors.Errorf("spawnhost-stop job '%s' could not be found", j.SpawnhostStopJobID))
```

Comment:
We don't remove the SpawnhostStopJobID so theortically we would just loop forever when we retry here right?

Reply:
Yes – changed to `AddError`!